File: csm/runner.py
"""Orchestration loop: pick pending items, run them, park on limits."""

import os
import time
from datetime import datetime, timedelta

from . import claude_runner, config, ledger, queuefile, report, sessions


def resolve_until(until_hhmm: str | None) -> datetime | None:
    if not until_hhmm:
        return None
    hour, minute = (int(p) for p in until_hhmm.split(":"))
    candidate = datetime.now().replace(hour=hour, minute=minute, second=0, microsecond=0)
    if candidate <= datetime.now():
        candidate += timedelta(days=1)
    return candidate


def run(until_hhmm: str | None = None, max_items: int | None = None,
        dry_run: bool = False, item_id: str | None = None) -> int:
    cfg = config.load()
    config.ensure_init()
    until = resolve_until(until_hhmm)
    registry = sessions.load()
    env, removed = claude_runner.stripped_env(dict(os.environ))
    if removed:
        print(f"csm: stripped billing-capable env vars from claude subprocess: {', '.join(removed)}")

    items = queuefile.load_items()
    if item_id:
        item, err = queuefile.find_item(items, item_id)
        if err:
            print(f"csm: {err}")
            return 1
        if item["status"] != queuefile.PENDING:
            print(f"csm: [{item['id']}] is {item['status']}, not pending "
                  f"(use `csm requeue {item['id']}` to run it again)")
            return 1
        todo = [item]
    else:
        todo = queuefile.pending_items(items)
    if not todo:
        print("csm: queue is empty - nothing to run")
        return 0

    if not dry_run:
        report.append_run_header("manual" if until is None else f"scheduled until {until:%H:%M}")

    ran = 0
    for item in todo:
        if until and datetime.now() >= until:
            print(f"csm: reached --until {until:%H:%M}, stopping")
            break
        if max_items is not None and ran >= max_items:
            print(f"csm: reached --max-items {max_items}, stopping")
            break

        spend = ledger.weekly_spend()
        if spend >= cfg["weekly_token_budget"]:
            msg = (f"weekly budget reached ({spend:,} / {cfg['weekly_token_budget']:,} "
                   "weighted tokens in the last 7 days)")
            print(f"csm: {msg}, stopping")
            if not dry_run:
                report.append_note(msg)
            break

        resume_id = None
        if not item.get("force_new_session"):
            resume_id = sessions.resumable_session(registry, cfg, item["project"])

        session = f"resume {resume_id}" if resume_id else "new session (context.md pickup)"
        model, effort, run_mode = claude_runner.item_settings(cfg, item)
        print(f"csm: [{item['id']}] {item['project']} | {session} | "
              f"model={model} effort={effort or 'cli-default'} mode={run_mode}")
        if dry_run:
            ran += 1
            continue

        result = claude_runner.run_item(cfg, item, resume_id, env)
        ran += 1

        if result.auth_error:
            print(f"csm: {result.error} - stopping the whole run (no quota burned on a broken login)")
            report.append_note(f"run aborted: {result.error}")
            return 2

        if result.rate_limited:
            print("csm: usage limit reached")
            report.append_note("usage limit reached"
                               + (f", resets ~{result.reset_at:%H:%M}" if result.reset_at else ""))
            if result.reset_at and until and result.reset_at < until:
                wait = (result.reset_at - datetime.now()).total_seconds() + 60
                print(f"csm: sleeping {wait / 60:.0f} min until reset (~{result.reset_at:%H:%M})")
                time.sleep(max(wait, 60))
                continue  # item stays pending, retry after reset
            print("csm: no reset time inside the run window - stopping")
            return 3

        _record_outcome(cfg, items, item, result, registry)

    print(f"csm: done ({ran} item(s) processed)")
    return 0


def _record_outcome(cfg, items, item, result, registry) -> None:
    if result.session_id:
        sessions.record(registry, item["project"], result.session_id, result.context_tokens)

    if result.usage:
        ledger.append(item["id"], item["project"],
                      item.get("model") or cfg["default_model"], result.usage)
    weighted = ledger.weighted(result.usage) if result.usage else 0
    spend = ledger.weekly_spend()

    status = queuefile.DONE if result.ok else queuefile.NEEDS_ATTENTION
    queuefile.finish_item(items, item, status, session_id=result.session_id,
                          summary=result.summary, error=result.error,
                          tokens=result.usage or None)
    report.append_item(item, status, result.session_id, result.summary,
                       result.error, weighted, spend, cfg["weekly_token_budget"])

    tag = "done" if result.ok else "NEEDS ATTENTION"
    detail = result.summary or result.error or ""
    print(f"csm: [{item['id']}] {tag} - {detail}")

File: internal/runner/runner.go
// Package runner picks pending items, runs them, and parks on limits.
package runner

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/EkinBarisC/claude-session-manager/internal/claude"
	"github.com/EkinBarisC/claude-session-manager/internal/config"
	"github.com/EkinBarisC/claude-session-manager/internal/ledger"
	"github.com/EkinBarisC/claude-session-manager/internal/queue"
	"github.com/EkinBarisC/claude-session-manager/internal/report"
	"github.com/EkinBarisC/claude-session-manager/internal/runstate"
	"github.com/EkinBarisC/claude-session-manager/internal/sessions"
)

// Exit codes for `csm run`.
const (
	ExitOK        = 0
	ExitUsage     = 1
	ExitAuthError = 2
	ExitRateLimit = 3
)

// ResolveUntil turns "HH:MM" into the next occurrence of that local time.
func ResolveUntil(hhmm string) (*time.Time, error) {
	if hhmm == "" {
		return nil, nil
	}
	t, err := time.Parse("15:04", hhmm)
	if err != nil {
		return nil, fmt.Errorf("invalid --until time %q (want HH:MM)", hhmm)
	}
	now := time.Now()
	candidate := time.Date(now.Year(), now.Month(), now.Day(), t.Hour(), t.Minute(), 0, 0, now.Location())
	if !candidate.After(now) {
		candidate = candidate.AddDate(0, 0, 1)
	}
	return &candidate, nil
}

// Run processes pending items until the queue, --until, --max-items, or the
// weekly budget stops it. itemID restricts the run to a single item.
func Run(untilHHMM string, maxItems int, dryRun bool, itemID string) int {
	cfg, err := config.Load()
	if err != nil {
		fmt.Println("csm:", err)
		return ExitUsage
	}
	config.EnsureInit()
	until, err := ResolveUntil(untilHHMM)
	if err != nil {
		fmt.Println("csm:", err)
		return ExitUsage
	}
	registry := sessions.Load()
	env, removed := claude.StrippedEnv(os.Environ())
	if len(removed) > 0 {
		fmt.Printf("csm: stripped billing-capable env vars from claude subprocess: %s\n", strings.Join(removed, ", "))
	}

	items := queue.Load()
	var todo []*queue.Item
	if itemID != "" {
		item, err := queue.Find(items, itemID)
		if err != nil {
			fmt.Println("csm:", err)
			return ExitUsage
		}
		if item.Status != queue.Pending {
			fmt.Printf("csm: [%s] is %s, not pending (use `csm requeue %s` to run it again)\n",
				item.ID, item.Status, item.ID)
			return ExitUsage
		}
		todo = []*queue.Item{item}
	} else {
		todo = queue.PendingItems(items)
	}
	if len(todo) == 0 {
		fmt.Println("csm: queue is empty - nothing to run")
		return ExitOK
	}

	trigger := "manual"
	if until != nil {
		trigger = "scheduled until " + until.Format("15:04")
	}

	ran := 0
	outcome := "completed"
	var held *runstate.Held
	// Dry runs don't invoke claude or mutate anything, so they take no
	// lock and leave no heartbeat.
	if !dryRun {
		held, err = runstate.Acquire(trigger)
		if err != nil {
			fmt.Println("csm:", err)
			return ExitUsage
		}
		defer held.Release()
		startedAt := time.Now().UTC().Format(time.RFC3339)
		defer func() {
			runstate.WriteLastRun(startedAt, trigger, ran, outcome)
		}()
		report.AppendRunHeader(trigger)
		// real plan usage (free to query) so the report shows what the
		// night started from
		if limits, _, err := claude.FetchUsage(cfg); err == nil && len(limits) > 0 {
			line := claude.FormatLimits(limits)
			fmt.Println("csm: plan usage: " + line)
			report.AppendNote("plan usage at start: " + line)
		}
	}

	for _, item := range todo {
		if until != nil && !time.Now().Before(*until) {
			fmt.Printf("csm: reached --until %s, stopping\n", until.Format("15:04"))
			outcome = "reached --until " + until.Format("15:04")
			break
		}
		if maxItems > 0 && ran >= maxItems {
			fmt.Printf("csm: reached --max-items %d, stopping\n", maxItems)
			outcome = fmt.Sprintf("reached --max-items %d", maxItems)
			break
		}

		spend := ledger.WeeklySpend()
		if spend >= cfg.WeeklyTokenBudget {
			msg := fmt.Sprintf("weekly budget reached (%d / %d weighted tokens in the last 7 days)",
				spend, cfg.WeeklyTokenBudget)
			fmt.Printf("csm: %s, stopping\n", msg)
			outcome = "weekly budget reached"
			if !dryRun {
				report.AppendNote(msg)
			}
			break
		}

		resumeID := ""
		if !item.ForceNewSession {
			resumeID = registry.Resumable(cfg, item.Project)
		}

		session := "new session (context.md pickup)"
		if resumeID != "" {
			session = "resume " + resumeID
		}
		model, effort, mode := claude.ItemSettings(cfg, item)
		if effort == "" {
			effort = "cli-default"
		}
		fmt.Printf("csm: [%s] %s | %s | model=%s effort=%s mode=%s\n",
			item.ID, item.Project, session, model, effort, mode)
		if dryRun {
			ran++
			continue
		}

		held.SetItem(item.ID)
		result := runItemWithProgress(cfg, item, resumeID, env)
		ran++

		if result.AuthError {
			fmt.Printf("csm: %s - stopping the whole run (no quota burned on a broken login)\n", result.Error)
			report.AppendNote("run aborted: " + result.Error)
			outcome = "aborted: " + result.Error
			return ExitAuthError
		}

		if result.RateLimited {
			fmt.Println("csm: usage limit reached")
			note := "usage limit reached"
			if result.ResetAt != nil {
				note += ", resets ~" + result.ResetAt.Format("15:04")
			}
			report.AppendNote(note)
			if result.ResetAt != nil && until != nil && result.ResetAt.Before(*until) {
				wait := time.Until(*result.ResetAt) + time.Minute
				if wait < time.Minute {
					wait = time.Minute
				}
				fmt.Printf("csm: sleeping %.0f min until reset (~%s)\n",
					wait.Minutes(), result.ResetAt.Format("15:04"))
				time.Sleep(wait)
				continue // item stays pending, retry after reset
			}
			fmt.Println("csm: no reset time inside the run window - stopping")
			outcome = "usage limit reached, no reset inside the run window"
			return ExitRateLimit
		}

		RecordOutcome(cfg, items, item, result, registry)

		tag := "done"
		if !result.OK {
			tag = "NEEDS ATTENTION"
		}
		detail := result.Summary
		if detail == "" {
			detail = result.Error
		}
		fmt.Printf("csm: [%s] %s - %s\n", item.ID, tag, detail)
	}

	fmt.Printf("csm: done (%d item(s) processed)\n", ran)
	return ExitOK
}

// runItemWithProgress runs one item while keeping the terminal alive: on a
// TTY a single line is redrawn with the elapsed time; when output is piped
// (cron log) a plain line is printed once a minute instead.
func runItemWithProgress(cfg config.Config, item *queue.Item, resumeID string, env []string) claude.Result {
	done := make(chan claude.Result, 1)
	go func() { done <- claude.RunItem(cfg, item, resumeID, env) }()

	tty := stdoutIsTTY()
	interval := time.Minute
	if tty {
		interval = 2 * time.Second
	}
	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	start := time.Now()
	for {
		select {
		case result := <-done:
			if tty {
				fmt.Printf("\r%*s\r", 70, "")
			}
			return result
		case <-ticker.C:
			elapsed := time.Since(start).Truncate(time.Second)
			if tty {
				fmt.Printf("\r  running [%s] for %s (timeout %dm) ", item.ID, elapsed, cfg.ItemTimeoutMinutes)
			} else {
				fmt.Printf("csm: [%s] still running after %s\n", item.ID, elapsed)
			}
		}
	}
}

func stdoutIsTTY() bool {
	info, err := os.Stdout.Stat()
	return err == nil && info.Mode()&os.ModeCharDevice != 0
}

// RecordOutcome persists everything one finished run produced: session
// registry, ledger, queue status, and the report block.
func RecordOutcome(cfg config.Config, items []*queue.Item, item *queue.Item,
	result claude.Result, registry sessions.Registry) {
	if result.SessionID != "" {
		registry.Record(item.Project, result.SessionID, result.ContextTokens())
	}

	model, _, _ := claude.ItemSettings(cfg, item)
	weighted := 0
	if len(result.Usage) > 0 {
		ledger.Append(item.ID, item.Project, model, result.Usage)
		weighted = ledger.Weighted(result.Usage)
	}
	spend := ledger.WeeklySpend()

	status := queue.Done
	if !result.OK {
		status = queue.NeedsAttention
	}
	queue.Finish(items, item, status, result.SessionID, result.Branch,
		result.Summary, result.Error, result.Usage)
	logPath := writeTranscript(item, result)
	report.AppendItem(item, status, result.SessionID, result.Summary, result.Error,
		weighted, spend, cfg.WeeklyTokenBudget, logPath)
}

// writeTranscript saves the run's full result text so failures can be
// diagnosed without resuming the session. Returns the file path, or ""
// when there was nothing to save.
func writeTranscript(item *queue.Item, result claude.Result) string {
	if strings.TrimSpace(result.ResultText) == "" {
		return ""
	}
	if err := os.MkdirAll(config.LogsDir(), 0o755); err != nil {
		return ""
	}
	path := filepath.Join(config.LogsDir(), item.ID+".md")
	header := fmt.Sprintf("# csm transcript [%s]\n\n- project: %s\n- finished: %s\n",
		item.ID, item.Project, item.FinishedAt)
	if item.Branch != "" {
		header += "- branch: " + item.Branch + "\n"
	}
	if result.Error != "" {
		header += "- error: " + result.Error + "\n"
	}
	if err := os.WriteFile(path, []byte(header+"\n---\n\n"+result.ResultText+"\n"), 0o644); err != nil {
		return ""
	}
	return path
}
